tools/data-handler/src/utils/clingo-facts.ts

```diff
@@ -0,0 +1,260 @@
+  for (const transition of workflow.transitions) {
+    for (const from of transition.fromState) {
+      builder.addFact(
+        'workflowTransition',
+        workflow.name,
+        transition.name,
+        from,
+        transition.toState,
+      );
+    }
+    if (transition.fromState.length === 0) {
+      builder.addFact(
+        'workflowTransition',
+        workflow.name,
+        transition.name,
+        '',
+        transition.toState,
+      );
+    }
+  }
+  return builder.buildAll();
+};
+
+export const createCardFacts = async (card: Card, project: Project) => {
+  // Small helper to deduce parent path
+  function parentPath(cardPath: string) {
+    const pathParts = cardPath.split(sep);
+    if (pathParts.at(pathParts.length - 2) === 'cardRoot') {
+      return '';
+    } else {
+      return pathParts.at(pathParts.length - 3);
+    }
+  }
+  const parentsPath = parentPath(card.path);
+  const builder = new ClingoProgramBuilder().addComment(card.key);
+
+  if (card.metadata) {
+    for (const [field, value] of Object.entries(card.metadata)) {
+      if (field === 'labels') {
+        for (const label of value as Array<string>) {
+          builder.addCustomFact('label', (b) =>
+            b.addLiteralArgument(card.key).addArgument(label),
+          );
+        }
+      } else if (field === 'links') {
+        for (const link of value as Link[]) {
+          builder.addCustomFact('link', (b) =>
+            b
+              .addLiteralArguments(card.key, link.cardKey)
+              .addArguments(link.linkType, link.linkDescription ?? null),
+          );
+        }
+      } else {
+        // Do not write null values
+        if (value == null) {
+          continue;
+        }
+        const fieldType = await project.fieldType(field);
+        builder.addCustomFact('field', (b) =>
+          b
+            .addLiteralArgument(card.key)
+            .addArguments(
+              field,
+              fieldType && ['datetime', 'date'].includes(fieldType.dataType)
+                ? new Date(value as string).getTime()
+                : value.toString(),
+            ),
+        );
+      }
+    }
+  }
+
+  if (parentsPath !== undefined && parentsPath !== '') {
+    builder.addCustomFact('parent', (b) =>
+      b.addLiteralArguments(card.key, parentsPath),
+    );
+  }
+
+  return builder.buildAll();
+};
+
+export const createFieldTypeFacts = (fieldType: FieldType) => {
+  const builder = new ClingoProgramBuilder();
+  builder.addFact('fieldType', fieldType.name);
+
+  if (fieldType.displayName)
+    builder.addFact(
+      'field',
+      fieldType.name,
+      'displayName',
+      fieldType.displayName,
+    );
+
+  if (fieldType.fieldDescription)
+    builder.addFact(
+      'field',
+      fieldType.name,
+      'fieldDescription',
+      fieldType.fieldDescription,
+    );
+
+  builder.addFact('field', fieldType.name, 'dataType', fieldType.dataType);
+
+  if (fieldType.enumValues) {
+    let index = 1;
```

Comment:
Clingo arguments are 1-based array? 
Or is there something "special" in the zero index?

Reply:
@henryhaverinen Is there a reason why we use 1-based indeces? I guess they are only used for ordering right now, so it cold technically be any number and still work just fine. But is there some other purpose for this?

Reply:
There is no reason in Clingo for the choice of 1-based indeces. I probably just first chose some kind of 1-based indexing in the query language for the select term: select(1, "title") for selecting on the top level of the results, and then used 1-based indexes in the next occurrence. We can change to 0-based indexes everywhere if you think that'd be better.

Reply:
This has no impact on functionality as of yet, so we can discuss this later, but I'll merge this ticket